docs/man/man1/juicer.1
'\" t
.\"     Title: juicer
.\"    Author: :doctype:manpage
.\" Generator: DocBook XSL Stylesheets v1.78.1 <http://docbook.sf.net/>
.\"      Date: 06/30/2013
.\"    Manual: Pulp repos and release carts
.\"    Source: Juicer 0.5.0
.\"  Language: English
.\"
.TH "JUICER" "1" "06/30/2013" "Juicer 0\&.5\&.0" "Pulp repos and release carts"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
juicer \- create and push release carts, upload items to pulp
.SH "SYNOPSIS"
.sp
juicer [\-vvv] subcommand [PARAMETERS \&...] [OPTIONS \&...]
.SH "DESCRIPTION"
.sp
A release cart is a description of what RPMs you want to push into a specific repository\&. A single cart can specify multiple repos and associated items\&. Release cart descriptions are saved so you can easily "promote" a cart to another environment\&.
.sp
RPMs (\fIitems\fR) need not be stored on your local machine\&. Juicer can handle remote resources as well\&. Anywhere an option argument is described as \fIitem\fR you can give the path to a local directory, or rpm, as well as an RPM available over HTTP\&. HTTP directory indexes work as well; the page is parsed and links to rpm files are added to your cart\&.
.SH "COMMON OPTIONS"
.PP
\fB\-v\fR, \fB\-\-verbose\fR
.RS 4
Increase the verbosity (up to 3x)\&. In usage specify
\fI\-v\fR\fBbefore\fR\fISUBCOMMAND\fR\&.
.RE
.SH "CREATE"
.sp
usage: juicer create \fICARTNAME\fR [\-f manifest] [\-r \fIREPONAME\fR \fIitems\fR \&... [ \-r \fIREPONAME\fR \fIitems\fR ]]
.sp
Create a cart with the items specified\&.
.PP
\fBCARTNAME\fR
.RS 4
The name of the new release cart\&.
.RE
.PP
\fB\-f\fR \fImanifest\fR
.RS 4
Create a cart from a manifest file\&. A manifest file is written in the following format:
.RE
.sp
name: version\-release
.PP
\fB\-r\fR \fIREPONAME\fR
.RS 4
Name of the reopsitory to install
\fIITEMS\fR
into\&. The
\fI\-r\fR
option may be given multiple times\&.
.RE
.PP
\fBITEM\fR \&...
.RS 4
Items to add to the cart in repository
\fIREPONAME\fR\&.
.RE
.SH "HELLO"
.sp
usage: juicer hello [\-\-in \fIENVIRONMENT\fR \&...]
.sp
Test connection settings in \fB~/\&.juicer\&.conf\fR
.PP
\fBENVIRONMENT\fR
.RS 4
The environments to limit connecting testing to\&.
.RE
.SH "LIST-CARTS"
.sp
usage: juicer list\-carts
.sp
List carts stored remotely\&.
.SH "PULL"
.sp
usage: juicer pull \fICARTNAME\fR
.sp
Pulls a description of a cart from the pulp server and saves it on your local machine in ~/\&.juicer\-carts/\&.
.PP
\fBCARTNAME\fR
.RS 4
The name of the cart to pull\&.
.RE
.SH "PUSH"
.sp
usage: juicer push \fICARTNAME\fR [\-\-in \fIENVIRONMENT\fR \&...]
.sp
Pushes/Updates a cart on the pulp server\&.
.PP
\fBCARTNAME\fR
.RS 4
The name of the cart to push\&.
.RE
.PP
\fB\-\-in\fR \fIENVIRONMENT\fR \&...
.RS 4
The environments to push the new/updated cart to\&.
.RE
.SH "PROMOTE"
.sp
usage: juicer promote \fICARTNAME\fR
.sp
Promote a release cart to the next release environment in its directed graph\&.
.PP
\fBCARTNAME\fR
.RS 4
The release cart to promote\&.
.RE
.SH "UPDATE"
.sp
usage: juicer update \fICARTNAME\fR [\-f \fIMANIFEST\fR] [\-r \fIREPONAME\fR \fIITEMS\fR \&... [ \-r \fIREPONAME\fR \fIitems\fR ]]
.sp
Update the specified cart with the items specified\&.
.sp
Updating a cart will add an item to the cart if it wasn\(cqt originally present or overwrite an item if it already was\&.
.PP
\fBCARTNAME\fR
.RS 4
The name of the new release cart\&.
.RE
.PP
\fB\-f\fR \fIMANIFEST\fR
.RS 4
Update a cart with a manifest file\&. A manifest file is written in the following format:
.RE
.sp
name: version\-release
.PP
\fB\-r\fR \fIREPONAME\fR
.RS 4
Name of the reopsitory to install
\fIITEMS\fR
into\&. The
\fI\-r\fR
option may be given multiple times\&.
.RE
.PP
\fBITEMS\fR \&...
.RS 4
Items to add to the cart in repository
\fIREPONAME\fR\&.
.RE
.SH "SEARCH"
.sp
usage: juicer search \fIITEM\fR [\-r \fIREPO\fR \&...] [\-c] [\-\-in \fIENVIRONMENT\fR \&...]
.sp
Search for an RPM (\fIitem\fR) in pulp\&.
.PP
\fBITEM\fR
.RS 4
Name of the RPM to search for\&.
.RE
.PP
\fB\-r\fR \fIREPO\fR \&...
.RS 4
The repo(s) to limit search scope to\&.
.RE
.PP
\fB\-c\fR
.RS 4
Search for the package in carts as well\&.
.RE
.PP
\fB\-\-in\fR \fIENVIRONMENT\fR \&...
.RS 4
The environments to limit search scope to\&.
.RE
.SH "SHOW"
.sp
usage: juicer show \fICARTNAME\fR
.sp
Print the contents of a cart\&.
.PP
\fBCARTNAME\fR
.RS 4
The name of the release cart to show\&.
.RE
.SH "UPLOAD"
.sp
usage: juicer upload \-r \fIREPO\fR \fIITEM\fR \&... [\-\-in \fIENVIRONMENT\fR \&...]
.sp
Upload multiple RPMs or files (\fIITEM\fR) to \fIREPO\fR\&.
.PP
\fB\-r\fR \fIREPO\fR \&...
.RS 4
The repo that
\fIITEM\fR
will be uploaded to\&. The
\fI\-r\fR
option may be given multiple times\&.
.RE
.PP
\fBITEM\fR \&...
.RS 4
Name of the RPM(s) or file(s) to upload\&.
.RE
.PP
\fB\-\-in\fR \fIENVIRONMENT\fR \&...
.RS 4
The environments which items will be uploaded to\&.
.RE
.SH "EXAMPLES"
.sp
\fIitems\fR given may be any number and combination of the following input resource types:
.PP
\fBlocal\fR
.RS 4
\&./directory/of/items | \&./path/to/item\&.rpm | /path/to/item*
.sp
.if n \{\
.RS 4
.\}
.nf
The items given are directly uploaded to the pulp server\&.
.fi
.if n \{\
.RE
.\}
.RE
.sp
Here is an example of how you would specify a directory of \fBlocal\fR RPMs and a direct path:
.sp
.if n \{\
.RS 4
.\}
.nf
$ juicer upload \-r javastuff \e
    ~/my\-java\-app/rpmbuild/noarch/ \e
    ~/misc\-java\-rpms/megafrobber\-0\&.8\&.0\&.noarch\&.rpm
.fi
.if n \{\
.RE
.\}
.PP
\fBremote\fR
.RS 4
http[s]://rpms/directory/ | http[s]://website/some\-thing\&.rpm
.sp
.if n \{\
.RS 4
.\}
.nf
URL to an HTTP directory index or an RPM\&. RPM paths are parsed
from the index and then added to your cart\&. All remote items are
synced when you upload or push\&.
.fi
.if n \{\
.RE
.\}
.RE
.sp
Here is an example of how you would specify a directory of \fBremote\fR RPMs and a specific remote RPM:
.sp
.if n \{\
.RS 4
.\}
.nf
$ juicer upload \-r javastuff \e
    http://jenkins\&.foo/job/results/ \e
    http://foo\&.bar/rpms/megafrobber\-0\&.8\&.0\&.noarch\&.rpm
.fi
.if n \{\
.RE
.\}
.sp
\fBcreate and show a cart\fR:
.sp
.if n \{\
.RS 4
.\}
.nf
$ juicer create juicer\-0\&.1\&.7 \e
    \-r juicer \e
    http://kojipkgs\&.fedoraproject\&.org/\&.\&.\&.\&./juicer\-0\&.1\&.7\-1\&.fc17\&.noarch\&.rpm \e
    \-r juicer\-deps \e
    \&./rpm\-build/noarch/juicer\-misc\-0\&.1\&.7\-1\&.fc17\&.noarch\&.rpm
Creating cart \*(Aqjuicer\-0\&.1\&.7\*(Aq\&.
Saved cart \*(Aqjuicer\-0\&.1\&.7\*(Aq\&.
JUICER
\-\-\-\-\-\-
http://kojipkgs\&.fedoraproject\&.org/\&.\&.\&.\&./juicer\-0\&.1\&.7\-1\&.fc17\&.noarch\&.rpm
.fi
.if n \{\
.RE
.\}
.sp
.if n \{\
.RS 4
.\}
.nf
JUICER\-DEPS
\-\-\-\-\-\-\-\-\-\-\-
\&./rpm\-build/noarch/juicer\-misc\-0\&.1\&.7\-1\&.fc17\&.noarch\&.rpm
.fi
.if n \{\
.RE
.\}
.sp
The cart description is saved into \fB~/\&.juicer\-carts/\fR as \fBjuicer\-0\&.1\&.7\&.json\fR\&. We could show it again simply:
.sp
.if n \{\
.RS 4
.\}
.nf
$ juicer show juicer\-0\&.1\&.7
JUICER
\-\-\-\-\-\-
http://kojipkgs\&.fedoraproject\&.org/\&.\&.\&.\&./juicer\-0\&.1\&.7\-1\&.fc17\&.noarch\&.rpm
.fi
.if n \{\
.RE
.\}
.sp
.if n \{\
.RS 4
.\}
.nf
JUICER\-DEPS
\-\-\-\-\-\-\-\-\-\-\-
\&./rpm\-build/noarch/juicer\-misc\-0\&.1\&.7\-1\&.fc17\&.noarch\&.rpm
.fi
.if n \{\
.RE
.\}
.sp
Remote items will be synced automatically when we push this cart to the repositories\&. Items synced are saved into \fB~/\&.juicer\-carts/\fR\fICART\-NAME\fR\fB\-remotes/\fR\&.
.sp
Similarly, when using the \fBupload\fR command, remotes are also synced\&.
.SH "FILES"
.sp
\fB~/\&.juicer\&.conf\fR \(em Juicer configuration file
.sp
\fB~/\&.juicer\-carts/\fR \(em Cart storage location
.SH "AUTHOR"
.sp
Juicer was written by GCA\-PC, Red Hat, Inc\&.\&.
.sp
This man page was written by Tim Bielawa <tbielawa@redhat\&.com>\&.
.SH "COPYRIGHT"
.sp
Copyright \(co 2012, Red Hat, Inc\&.\&.
.sp
Juicer is released under the terms of the GPLv3+ License\&.
.SH "SEE ALSO"
.sp
\fBjuicer\-admin\fR(1), \fBjuicer\&.conf\fR(5)
.sp
The Juicer Homepage: https://github\&.com/juicer/juicer/
.SH "AUTHOR"
.PP
\fB:doctype:manpage\fR
.RS 4
Author.
.RE


juicer/juicer/Juicer.py
# -*- coding: utf-8 -*-
# Juicer - Administer Pulp and Release Carts
# Copyright © 2012,2013, Red Hat, Inc.
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program. If not, see <http://www.gnu.org/licenses/>.

from juicer.common import Constants
from juicer.common.Errors import *
import juicer.common.Cart
import juicer.juicer
import juicer.utils
import juicer.utils.Upload
import os
import re


class Juicer(object):
    def __init__(self, args):
        self.args = args

        (self.connectors, self._defaults) = juicer.utils.get_login_info()

        if 'environment' in self.args:
            for env in self.args.environment:
                if env not in self.connectors.keys():
                    raise JuicerKeyError("%s is not an environment defined in juicer.conf" % env)

    # this is used to upload carts to pulp
    def upload(self, env, cart):
        """
        Nothing special happens here. This method recieves a
        destination repo, and a payload of `cart` which will be
        uploaded into the target repo.

        Preparation: To use this method you must pre-process your
        cart: Remotes must be fetched and saved locally. Directories
        must be recursed and replaced with their contents. Items
        should be signed if necessary.

        Warning: this method trusts you to Do The Right Thing (TM),
        ahead of time and check file types before feeding them to it.

        `env` - name of the environment with the cart destination
        `cart` - cart to upload
        """
        for repo in cart.repos():
            if not juicer.utils.repo_exists_p(repo, self.connectors[env], env):
                juicer.utils.Log.log_info("repo '%s' doesn't exist in %s environment... skipping!",
                                          (repo, env))
                continue

            repoid = "%s-%s" % (repo, env)
            juicer.utils.Log.log_debug("Beginning upload into %s repo" % repoid)

            for item in cart[repo]:
                juicer.utils.Log.log_info("Initiating upload of '%s' into '%s'" % (item.path, repoid))
                rpm_id = juicer.utils.upload_rpm(item.path, repoid, self.connectors[env])
                juicer.utils.Log.log_debug('%s uploaded with an id of %s' %
                                           (os.path.basename(item.path), rpm_id))

            self.connectors[env].post('/repositories/%s/actions/publish/' % repoid, {'id': 'yum_distributor'})

            # Upload carts aren't special, don't update their paths
            if cart.cart_name == 'upload-cart':
                continue

            # Set the path to items in this cart to their location on
            # the pulp server.
            for item in cart[repo]:
                path = juicer.utils.remote_url(self.connectors[env],
                                               env,
                                               repo,
                                               os.path.basename(item.path))
                item.update(path)

        # Upload carts don't persist
        if not cart.cart_name == 'upload-cart':
            cart.save()
            self.publish(cart)
        return True

    def push(self, cart, env=None):
        """
        `cart` - Release cart to push items from

        Pushes the items in a release cart to the pre-release environment.
        """
        juicer.utils.Log.log_debug("Initializing push of cart '%s'" % cart.cart_name)

        if not env:
            env = self._defaults['start_in']

        self.sign_cart_for_env_maybe(cart, env)
        self.upload(env, cart)
        return True

    def publish(self, cart, env=None):
        """
        `cart` - Release cart to publish in json format

        Publish a release cart in JSON format to the pre-release environment.
        """
        juicer.utils.Log.log_debug("Initializing publish of cart '%s'" % cart.cart_name)

        if not env:
            env = self._defaults['start_in']

        cart_id = juicer.utils.upload_cart(cart, env)
        juicer.utils.Log.log_debug('%s uploaded with an id of %s' %
                                   (cart.cart_name, cart_id))
        return True

    def create(self, cart_name, cart_description):
        """
        `cart_name` - Name of this release cart
        `cart_description` - list of ['reponame', item1, ..., itemN] lists
        """
        cart = juicer.common.Cart.Cart(cart_name)

        # repo_items is a list that starts with the REPO name,
        # followed by the ITEMS going into the repo.
        for repo_items in cart_description:
            (repo, items) = (repo_items[0], repo_items[1:])
            juicer.utils.Log.log_debug("Processing %s input items for repo '%s'." % (len(items), repo))
            cart[repo] = items

        cart.save()
        return cart

    def update(self, cart_name, cart_description, manifests):
        """
        `cart_name` - Name of this release cart
        `cart_description` - list of ['reponame', item1, ..., itemN] lists
        `manifests` - a list of manifest files
        """
        if cart_description is None:
            juicer.utils.Log.log_debug("No cart_description provided.")
            cart_description = []

        if manifests is None:
            juicer.utils.Log.log_debug("No manifests provided.")
            manifests = []

        juicer.utils.Log.log_debug("Loading cart '%s'." % cart_name)
        cart = juicer.common.Cart.Cart(cart_name, autoload=True)

        for repo_items in cart_description:
            (repo, items) = (repo_items[0], repo_items[1:])
            juicer.utils.Log.log_debug("Processing %s input items for repo '%s'." % (len(items), repo))

            for item in items:
                cart[repo].append(juicer.common.CartItem.CartItem(os.path.expanduser(item)))

        for manifest in manifests:
            cart.add_from_manifest(manifest, self.connectors)

        cart.save()
        return cart

    def create_manifest(self, cart_name, manifests):
        """
        `cart_name` - Name of this release cart
        `manifests` - a list of manifest files
        """
        cart = juicer.common.Cart.Cart(cart_name)

        for manifest in manifests:
            cart.add_from_manifest(manifest, self.connectors)

        cart.save()
        return cart

    def show(self, cart_name):
        cart = juicer.common.Cart.Cart(cart_name)
        cart.load(cart_name)
        return str(cart)

    def search(self, pkg_name=None, search_carts=False, query='/content/units/rpm/search/'):
        """
        search for a package stored in a pulp repo

        `pkg_name` - substring in the name of the package
        `search_carts` - whether or not to return carts that include
            the listed package
        """
        # this data block is... yeah. searching in pulp v2 is painful
        #
        # https://pulp-dev-guide.readthedocs.org/en/latest/rest-api/content/retrieval.html#search-for-units
        # https://pulp-dev-guide.readthedocs.org/en/latest/rest-api/conventions/criteria.html#search-criteria
        #
        # those are the API docs for searching
        data = {
                    'criteria': {
                        'filters': {'name': {'$regex': ".*%s.*" % pkg_name}},
                        'sort': [['name', 'ascending']],
                        'fields': ['name', 'description', 'version', 'release', 'arch', 'filename']
                    },
                    'include_repos': 'true'
                }
        repos = []

        juicer.utils.Log.log_info('Packages:')

        for env in self.args.environment:
            juicer.utils.Log.log_debug("Querying %s server" % env)
            _r = self.connectors[env].post(query, data)

            if not _r.status_code == Constants.PULP_POST_OK:
                juicer.utils.Log.log_debug("Expected PULP_POST_OK, got %s", _r.status_code)
                _r.raise_for_status()

            juicer.utils.Log.log_info('%s:' % str.upper(env))

            pkg_list = juicer.utils.load_json_str(_r.content)

            for package in pkg_list:
                # if the package is in a repo, show a link to the package in said repo
                # otherwise, show nothing
                if len(package['repository_memberships']) > 0:
                    target = package['repository_memberships'][0]

                    _r = self.connectors[env].get('/repositories/%s/' % target)
                    if not _r.status_code == Constants.PULP_GET_OK:
                        raise JuicerPulpError("%s was not found as a repoid. A %s status code was returned" %
                                (target, _r.status_code))
                    repo = juicer.utils.load_json_str(_r.content)['display_name']
                    repos.append(repo)

                    link = juicer.utils.remote_url(self.connectors[env], env, repo, package['filename'])
                else:
                    link = ''

                juicer.utils.Log.log_info('%s\t%s\t%s\t%s' % (package['name'], package['version'], package['release'], link))

        if search_carts:
            # if the package is in a cart, show the cart name
            juicer.utils.Log.log_info('\nCarts:')

            for env in self.args.environment:
                carts = juicer.utils.search_carts(env, pkg_name, repos)
                for cart in carts:
                    juicer.utils.Log.log_info(cart['_id'])

    def hello(self):
        """
        Test pulp server connections defined in ~/.juicer.conf.
        """
        for env in self.args.environment:
            juicer.utils.Log.log_info("Trying to open a connection to %s, %s ...",
                                      env, self.connectors[env].base_url)
            try:
                _r = self.connectors[env].get()
                juicer.utils.Log.log_info("OK")
            except JuicerError:
                juicer.utils.Log.log_info("FAILED")
                continue

            juicer.utils.Log.log_info("Attempting to authenticate as %s",
                                      self.connectors[env].auth[0])

            _r = self.connectors[env].get('/repositories/')

            if _r.status_code == Constants.PULP_GET_OK:
                juicer.utils.Log.log_info("OK")
            else:
                juicer.utils.Log.log_info("FAILED")
                juicer.utils.Log.log_info("Server said: %s", _r.content)
                continue
        return True

    def pull(self, cartname=None, env=None):
        """
        `cartname` - Name of cart

        Pull remote cart from the pre release (base) environment
        """
        if not env:
            env = self._defaults['start_in']
        juicer.utils.Log.log_debug("Initializing pulling cart: %s ...", cartname)

        cart_file = os.path.join(juicer.common.Cart.CART_LOCATION, cartname)
        cart_file += '.json'

        juicer.utils.write_json_document(cart_file, juicer.utils.download_cart(cartname, env))

        return True

    def promote(self, cart_name):
        """
        `name` - name of cart

        Promote a cart from its current environment to the next in the chain.
        """
        cart = juicer.common.Cart.Cart(cart_name=cart_name, autoload=True, autosync=True)
        old_env = cart.current_env
        cart.current_env = juicer.utils.get_next_environment(cart.current_env)

        # figure out what needs to be done to promote packages
        if juicer.utils.env_same_host(old_env, cart.current_env) and (self.connectors[old_env].requires_signature == self.connectors[cart.current_env].requires_signature):
            # iterate through packages and associate to new repo
            for repo, items in cart.iterrepos():
                query = '/repositories/%s/actions/associate/' % repo
                regex = re.compile('(.*)-(.*)')

                for item in items:
                    new_repo = ''
                    _m = regex.match(repo)
                    if _m:
                        new_repo = '%s-%s' % (_m.group(0), cart.current_env)

                    data = {
                            'source_repo_id': new_repo,
                            'criteria': {
                                'type_ids': ['rpm'],
                                'filters': {
                                    'unit': {
                                        '$and': [{'name': item.name}, {'version': item.version}, {'release': item.release}]
                                        }
                                    }
                                }
                            }
                    self.connectors[cart.current_env].post(query, data)
        else:
            juicer.utils.Log.log_debug("Syncing down rpms...")
            cart.sync_remotes()
            self.sign_cart_for_env_maybe(cart, cart.current_env)

            juicer.utils.Log.log_info("Promoting %s from %s to %s" %
                    (cart_name, old_env, cart.current_env))

            for repo in cart.repos():
                juicer.utils.Log.log_debug("Promoting %s to %s in %s" %
                                           (cart[repo], repo, cart.current_env))
                self.upload(cart.current_env, cart)

        cart.save()

        self.publish(cart)

    def list_carts(self):
        """
        List carts stored remotely.
        """
        for cart in sorted(juicer.utils.list_carts(self._defaults['environments'][0]),
                           key=lambda cart: cart['_id']):
            juicer.utils.Log.log_info("%s", cart['_id'])
        return True

    def sign_cart_for_env_maybe(self, cart, env=None):
        """
        Sign the items to upload, if the env requires a signature.

        `cart` - Cart to sign
        `envs` - The cart is signed if env has the property:
        requires_signature = True

        Will attempt to load the rpm_sign_plugin defined in
        ~/.juicer.conf, which must be a plugin inheriting from
        juicer.common.RpmSignPlugin. If available, we'll call
        cart.sign_items() with a reference to the
        rpm_sign_plugin.sign_rpms method.
        """
        if self.connectors[env].requires_signature:
            juicer.utils.Log.log_notice("%s requires RPM signatures", env)
            juicer.utils.Log.log_notice("Checking for rpm_sign_plugin definition ...")
            module_name = self._defaults['rpm_sign_plugin']
            if self._defaults['rpm_sign_plugin']:
                juicer.utils.Log.log_notice("Found rpm_sign_plugin definition: %s",
                                            self._defaults['rpm_sign_plugin'])
                juicer.utils.Log.log_notice("Attempting to load ...")

                try:
                    rpm_sign_plugin = __import__(module_name, fromlist=[module_name])
                    juicer.utils.Log.log_notice("Successfully loaded %s ...", module_name)
                    plugin_object = getattr(rpm_sign_plugin, module_name.split('.')[-1])
                    signer = plugin_object()
                    cart.sign_items(signer.sign_rpms)
                except ImportError as e:
                    juicer.utils.Log.log_notice("there was a problem using %s ... error: %s",
                                                module_name, e)

                if not juicer.utils.rpms_signed_p([item.path for item in cart.items()]):
                    raise JuicerNotSignedError('RPMs have not been signed.')

            else:
                raise JuicerConfigError("Did not find an rpm_sign_plugin in config file.")
            return True
        else:
            return None
